### PR TITLE
Add and modify mappings between version ids and release versions

### PR DESCRIPTION
**Pull request by thecoop**
Add a `ReleaseVersions` class to map between ids and release versions in error messages/logs, and gradle tasks to extract current version information and apply it to the version records in the repo

### CODE REVIEW

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/release/ReleaseToolsPlugin.java [hunk 1]
```diff
@@ -47,6 +47,9 @@ public void apply(Project project) {
 
         final Version version = VersionProperties.getElasticsearchVersion();
 
+        project.getTasks()
```
Comment (mark-vieira) [inline build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/release/ReleaseToolsPlugin.java:50]: Are these changes intended to be included in this PR?

Comment (thecoop) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/release/ReleaseToolsPlugin.java]: Not now #103335 has been merged (this branch is based on that one)

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/release/TagVersionsTask.java [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+
+public class TagVersionsTask extends DefaultTask {
```
Comment (mark-vieira) [inline build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/release/TagVersionsTask.java:37]: Am I missing something? I don't see where this task is actually used/created.

File: server/src/main/java/org/elasticsearch/ReleaseVersions.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+            return Collections.emptyNavigableMap();
+        }
+    }
+
+    private static NavigableMap<Integer, String> findVersionsMap(Class<?> versionContainer) {
+        return RESOLVED_VERSIONS.computeIfAbsent(versionContainer, ReleaseVersions::readVersionsFile);
+    }
+
+    public static String findReleaseVersion(Class<?> versionContainer, int id) {
+        if (USES_VERSIONS == false) return Integer.toString(id);
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/ReleaseVersions.java:94]: I think this means in serverless we would print the raw id. Is that what we want? The id is just as useless to a user there. Are we assuming a serverless user should never hit an error that would need to print the release version since it is kept up to date?

File: server/src/main/java/org/elasticsearch/ReleaseVersions.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+        // go for a range
+        if (versions.isEmpty()) {
+            // uh oh, something went wrong reading the file
+            return "<" + id + ">";
+        }
+
+        StringBuilder range = new StringBuilder();
+
+        var lowerRange = versions.lowerEntry(id);
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/ReleaseVersions.java:110]: If we fall down to here, something is wrong. On prem users would only ever be using released versions. There shouldn't be any uses of eg transport versions between on-prem releases. I think we could simplify this and fallback to just the id as above in the "the version map is empty" case (which also should never happen...maybe let's add an assert for this case?)

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/release/TagVersionsTask.java [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+        }
+        return extractor.highestVersionId;
+    }
+
+    private static class FieldIdExtractor implements Consumer<FieldDeclaration> {
+        private Integer highestVersionId;
+
+        @Override
+        public void accept(FieldDeclaration fieldDeclaration) {
+            var ints = fieldDeclaration.findAll(IntegerLiteralExpr.class);
```
Comment (rjernst) [inline build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/release/TagVersionsTask.java:123]: This seems fragile, wouldn't any other integer literal get caught up by this?

**Review by rjernst — COMMENTED** [review]
This will be a nice addition to give users back versions they know about! I have a few thoughts.

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/release/TagVersionsTask.java (hunk 2 above)
Comment (thecoop) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/release/TagVersionsTask.java:37]: No, this needs corresponding changes in release-tools (this PR is still WIP)

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/release/TagVersionsTask.java [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+                    }
+                }
+                default -> LOGGER.warn("Multiple integers found in version field declaration [{}]", fieldDeclaration); // and ignore it
+            }
+        }
+    }
+
+    private static final Pattern VERSION_LINE = Pattern.compile("(\\d+\\.\\d+\\.\\d+),(\\d+)(\\h*#.*)?");
+
+    static boolean addVersionRecord(List<String> versionRecordLines, Version release, int id) {
```
[the remark this answers is not in the document]
Comment (thecoop) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/release/TagVersionsTask.java:143]: We want to maintain any comments in the files too, which means keeping the collection in file order and modifying round those

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/release/TagVersionsTask.java [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+            versionRecordLines.add(0, fileRecord);
+        } else if (firstReleaseGreaterThan == -1) {
+            // no release greater than this one, put it at the end
+            versionRecordLines.add(fileRecord);
+        } else if (lastReleaseLessThan < firstReleaseGreaterThan) {
+            // put it here, in the middle, right after the release less than (any comments here are likely for the line below)
+            versionRecordLines.add(lastReleaseLessThan + 1, fileRecord);
+        } else {
+            // the file is out of order. Put it at the end and complain a little
+            LOGGER.warn("Versions record file is out of order. It may need manually sorting into order.");
```
[the remark this answers is not in the document]
Comment (thecoop) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/release/TagVersionsTask.java:199]: Automatically sorting it may break any comments

File: server/src/main/java/org/elasticsearch/ReleaseVersions.java [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    private static final Map<Class<?>, NavigableMap<Integer, String>> RESOLVED_VERSIONS = new ConcurrentHashMap<>();
+
+    private static final Pattern COMMENT_LINE = Pattern.compile("^\\h*#");
+    private static final Pattern VERSION_LINE = Pattern.compile("(\\d+\\.\\d+\\.\\d+),(\\d+)(\\h*#.*)?");
+
+    static NavigableMap<Integer, String> readVersionsFile(Class<?> versionContainer) {
+        String versionsFileName = versionContainer.getSimpleName() + ".csv";
+        URL versionsFile = versionContainer.getResource(versionsFileName);
```
[the remark this answers is not in the document]
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/ReleaseVersions.java:47]: `getResourceAsStream` doesn't give us the resolved URL, which is useful for displaying in logging messages for clarity

File: server/src/main/java/org/elasticsearch/ReleaseVersions.java [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+            )
+        ) {
+            NavigableMap<Integer, String> versions = new TreeMap<>();
+            String line;
+            while ((line = reader.readLine()) != null) {
+                if (COMMENT_LINE.matcher(line).find()) continue; // a comment
+
+                var matcher = VERSION_LINE.matcher(line);
+                if (matcher.matches() == false) {
+                    Log.warn("Incorrect format for line [{}] in [{}]", line, versionsFile);
```
[the remark this answers is not in the document]
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/ReleaseVersions.java:65]: Good point - changed

File: server/src/main/java/org/elasticsearch/ReleaseVersions.java [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+                } catch (NumberFormatException e) {
+                    // cannot happen??? regex is wrong...
+                    assert false : "Regex allowed non-integer id through: " + e;
+                    Log.warn("Incorrect format for line [{}] in [{}]", line, versionsFile, e);
+                }
+            }
+
+            return Collections.unmodifiableNavigableMap(versions);
+        } catch (Exception e) {
+            Log.error("Could not read versions file [{}]", versionsFile, e);
```
[the remark this answers is not in the document]
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/ReleaseVersions.java:84]: At the moment, this is only first scanned when it is requested, so this would cause failures at random points - hence why it fails in non-fatal ways.

We could have it eagerly load all the version mappings on startup, then it could fail outright if the files are incorrect. That would give more predictability

File: server/src/main/java/org/elasticsearch/ReleaseVersions.java (hunk 3 above)
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/ReleaseVersions.java:94]: This is mostly aimed at log messages - where transport/index versions do have meaning. If we ever need to print out anything for serverless, it should be the raw id, not the release version.

File: server/src/main/java/org/elasticsearch/ReleaseVersions.java (hunk 4 above)
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/ReleaseVersions.java:110]: I'm largely thinking of snapshot builds where we don't have a release version, but its a valid transport/index version. We could add an assert for the lower range, but we still need to do something if this ever happens, somehow, in production. We shouldn't blow up because an error message can't be resolved properly.

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/release/TagVersionsTask.java (hunk 5 above)
Comment (thecoop) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/release/TagVersionsTask.java:123]: we do only have single int initializers in version classes. And it ignores those that have multiple ints in the initializer.

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/release/TagVersionsTask.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (thecoop) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/release/TagVersionsTask.java:143]: Actually, thinking about it, I'm not sure we need any comments in here - any comments on weird releases & patches can be put in the respective java files.

[the remark this answers is not in the document]
Comment (thecoop) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/release/TagVersionsTask.java]: I've taken out the ability to have comments, allowing us to use a sorted map

File: server/src/main/java/org/elasticsearch/ReleaseVersions.java [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+                if (matcher.matches() == false) {
+                    Log.warn("Incorrect format for line [{}] in [{}]", line, versionsFile);
+                    continue;
+                }
+                try {
+                    Integer id = Integer.valueOf(matcher.group(2));
+                    String version = matcher.group(1);
+                    String existing = versions.putIfAbsent(id, version);
+                    if (existing != null) {
+                        Log.warn("Duplicate id [{}] for versions [{}, {}] in [{}]", id, existing, version, versionsFile);
```
[the remark this answers is not in the document]
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/ReleaseVersions.java]: It now blows up completely

File: server/src/main/java/org/elasticsearch/ReleaseVersions.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/ReleaseVersions.java]: I've change it to do this - eagerly load the files on startup, and so blowing up if the file is malformed

File: server/src/main/java/org/elasticsearch/ReleaseVersions.java (hunk 4 above)
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/ReleaseVersions.java]: I've added an assert in the rewritten code

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/release/TagVersionsTask.java (hunk 2 above)
Comment (thecoop) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/release/TagVersionsTask.java:41]: I've created a PR for the release tools to use these new tasks

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/release/ExtractCurrentVersionsTask.java [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+        int transportVersion = readLatestVersion(rootDir.resolve(TRANSPORT_VERSION_FILE_PATH));
+        LOGGER.lifecycle("Transport version: {}", transportVersion);
+        output.add(TRANSPORT_VERSION_TYPE + ":" + transportVersion);
+
+        int indexVersion = readLatestVersion(rootDir.resolve(INDEX_VERSION_FILE_PATH));
+        LOGGER.lifecycle("Index version: {}", indexVersion);
+        output.add(INDEX_VERSION_TYPE + ":" + indexVersion);
+
+        LOGGER.lifecycle("Writing version information to {}", outputFile);
+        Files.write(outputFile, output, StandardOpenOption.WRITE, StandardOpenOption.TRUNCATE_EXISTING);
```
Comment (mark-vieira) [inline build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/release/ExtractCurrentVersionsTask.java:65]: We need to add the `CREATE` option here otherwise this task fails when this file doesn't already exist.

**Review by mark-vieira — CHANGES_REQUESTED** [review]

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/release/ExtractCurrentVersionsTask.java (hunk 12 above)
Comment (thecoop) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/release/ExtractCurrentVersionsTask.java]: Ah yes, the release automation already creates the file, but it should work anyway

**Comment by thecoop** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/ReleaseVersions.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (rjernst) [reply server/src/main/java/org/elasticsearch/ReleaseVersions.java]: Shouldn't the versionsFileName be sufficient for a log message?

File: server/src/main/java/org/elasticsearch/ReleaseVersions.java [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+                    lowerBound = "[" + id + "]";
+                }
+
+                var upperRange = versions.higherEntry(id);
+                if (upperRange != null) {
+                    // too hard to guess what version this id might be for using the next version - just use it directly
+                    upperBound = upperRange.getValue().first().toString();
+                } else {
+                    // likely a version used after this code was released - ok
+                    upperBound = "[" + id + "]";
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/ReleaseVersions.java:108]: Could this be an assert just like above?

File: server/src/test/java/org/elasticsearch/ReleaseVersionsTests.java [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+        assertThat(versions.apply(21), equalTo("8.2.0"));
+        assertThat(versions.apply(22), equalTo("8.2.1"));
+    }
+
+    public void testReturnsRange() {
+        IntFunction<String> versions = ReleaseVersions.generateVersionsLookup(ReleaseVersionsTests.class);
+
+        assertThat(versions.apply(17), equalTo("8.1.2-8.2.0"));
+        expectThrows(AssertionError.class, () -> versions.apply(9));
+        assertThat(versions.apply(24), equalTo("8.2.2-[24]"));
```
Comment (rjernst) [inline server/src/test/java/org/elasticsearch/ReleaseVersionsTests.java:33]: This would be confusing in a log message. Could we change this to omit the upper bound, ie `8.2.2-`?

**Review by rjernst — APPROVED** [review]
Looks good, a couple last thoughts, and Mark's comment on write options should be addressed as well.

File: server/src/test/java/org/elasticsearch/ReleaseVersionsTests.java (hunk 14 above)
Comment (thecoop) [reply server/src/test/java/org/elasticsearch/ReleaseVersionsTests.java:33]: I think just having a trailing `-` would be a bit confusing. This should only be seen in snapshot builds used by devs, so having the id displayed is useful for diagnose the problem the log/exception message is for. Is there a different format you think we could use?

Comment (ldematte) [reply server/src/test/java/org/elasticsearch/ReleaseVersionsTests.java:33]: If this is only seen by "us" (devs) I think it is useful for debugging

**Review by ldematte — COMMENTED** [review]
I like it, and with the helpers in the XxxVersions classes it will be easy to use too.
I'm wondering if we should have something similar for features too - probably worth a separate discussion (and definitely a separate PR in case).

File: server/src/test/java/org/elasticsearch/ReleaseVersionsTests.java (hunk 14 above)
Comment (rjernst) [reply server/src/test/java/org/elasticsearch/ReleaseVersionsTests.java:33]: If it's only for devs, then perhaps being more explicit? I don't think devs will immediately understand this bracket notation (which will never be seen in production) is special and is a "raw" value.

Comment (thecoop) [reply server/src/test/java/org/elasticsearch/ReleaseVersionsTests.java:33]: How about something like `8.12.0-snapshot[8546000]`?

**Review by mark-vieira — APPROVED** [review]

File: server/src/test/java/org/elasticsearch/ReleaseVersionsTests.java (hunk 14 above)
Comment (rjernst) [reply server/src/test/java/org/elasticsearch/ReleaseVersionsTests.java:33]: That sounds good to me.
